20250429-RS-scheduler-broadcast-job-type.md

```diff
@@ -0,0 +1,320 @@
+# Scheduler Broadcast Job Type
+
+* Author(s): Josh van Leeuwen (@joshvanl)
+* Updated: 2025-04-29
+
+## Overview
+
+This proposal describes adding a new Scheduler Job type - `Broadcast`.
+This job type is designed to leverage the durability of Scheduler and its reliable connection to all daprd runtimes to broadcast objects to all relevant clients to allow for durable message broadcasting.
+The initial design focuses on two initial use cases, `actor pubsub` and `durable actor IDs`, but this job type's functionality can be extended to other domains in future.
+
+## Related Items
+
+https://github.com/dapr/proposals/pull/78
+
+## Background
+
+It is the case in Dapr that some features desire or require the need to be able to durably broadcast messages to some set of runtimes in the cluster.
+Such examples include the need to broadcast [Actor pubsub subscription](https://github.com/dapr/proposals/pull/78) manifests to all actor type implementations, or signal that an actor ID should be invoked, and life cycle extend beyond the typical actor deactivation timeout or dissemination events.
+It is imperative that these objects/messages are preserved across restarts/dissemination/crash events to ensure their durability.
+It is also the case that these messages need a reliable way to be broadcast to all the relevant hosts.
+
+While these two feature do have access to reliable durable storage in the form of the Actor State Store, implementing these require durable service discovery and co-ordination _at the runtime level_.
+Not to mention that storage is often not available outside of the context of Actors.
+Implementing these at the runtime level would be brittle as the co-ordination mechanism itself would require significant machinery.
+The Scheduler service already has application typed context service discovery and connectivity with durable storage access, so is the perfect place to implement this functionality.
+
+Users today are required to use "normal" Scheduler jobs to ensure message durable execution, and implement their own broadcast or message routing functionality.
+Typically creating a job which has an infinite schedule that triggers at a fixed small interval.
+Moving this functionality as a first concept to Dapr improves performance and improves the developer experience in creating complex distributed systems.
```

Comment:
How does this improve performance compared to "normal" Scheduler jobs to ensure message durable execution"?

Reply:
My take is that today we have "normal" Scheduler jobs for jobs and reminders. This gives us a mechanism to extend the concept to durable scheduled activities (as necessary for actor pub/sub and likely other use cases). I don't know it necessarily improves performance so much as abstract away the downstream purpose from the scheduler itself.

Reply:
It improves performance because today, one would have to rely on regular triggering of a job to signal a "keep alive" for the object to exist, without any easy way to signal a message has been deleted. Each trigger needs a round trip through the scheduler queue and call to each client over the network.

 This broadcast job type achieves CRUD message behaviour with the least number of Scheduler queue events and network calls .